rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,121 @@
+---
+feature: inherit-as-list
+start-date: 2021-10-17
+author: Ryan Burns (@r-burns)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: @synthetica9, @infinisil, @kevincox 
+shepherd-leader: @kevincox 
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `<attrset>.[ <attrnames> ]`,
+which constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+attrs.[ a b c ]
```

Comment:
One concern of this form is that `foo.[bar]` looks a lot like indexing operators in most languages. However I think this is a minor concern because the `.[` makes it stand out and this would rarely be used for single items.

Reply:
since the `|` operator is not a valid attribute or variable name, perhaps we could make use of it do to something like `[ with foo | a b c ]` instead?

Reply:
This syntax looks more foreign to me. It may also cause confusion with other languages that use `|` for bitwise or. I think I prefer the proposed solution.

Reply:
I think a lot of people familiar with functional programming are familiar with list comprehensions, this isn't exactly a list comprehension but the idea was attempting to use that familiarity to make it obvious what's happening. I guess it is a bit awkward after all though :shrug:

Reply:
I'd also suggest it introduces a small inconsistency in the syntax where a beginner now expects

`attrs.{ a b c }` (or `{ attrs | a b c }`)

to work the same way for attributes, but instead they need to learn `{ inherit (attrs) a b c; }` which appears arbitrarily different